src/values/AbstractObjectValue.js

```diff
@@ -561,6 +573,9 @@ export default class AbstractObjectValue extends AbstractValue {
       }
       invariant(d1val instanceof Value);
       invariant(d2val instanceof Value);
+      if (d1val === d2val) {
```

Comment:
This looked like an obvious optimization here, unless I'm mistaken?

Reply:
Yes, but joinValuesAsConditional should also do this optimization, so while this is faster, it is also a teeny bit of code bloat.

Reply:
It seems like something we should try to fix in joinValuesAsConditional. Any chance you can come up with a repro?